README.md
**RTFM** will save you a few precious keystrokes and clicks every time you need to "Read The... Flippin' Manual." Highlight a language keyword in the editor and RTFM will direct your browser search to your favourite reference resource for that term.

Which website you are taken to is determined by the document's current syntax (at bottom right corner of the editor window) and your preferred resource for that syntax (specified in the preferences). If no text is selected when RTFM is triggered (by menu or assigned hot-key), you will be prompted to enter a search string.

![Manual Search Entry](https://ext.runcode.run/rtfm/readme/RTFM_manual_entry.png)

RTFM includes settings for every syntax recognized by Nova, so even if you're working in
plain-text you could still do a general web search on any string, or set up a Custom URL to search the API of whatever you might be working on / documenting.

Lastly, if you're (say) working on SQL in a PHP file, you could temporarily switch the document's syntax to SQL so you can target your RTFM queries to the SQL resources instead of PHP.

*NOTE:* This extension is just a shortcut for performing the searches you might do on these sites or through search engines. If the search results aren't better when you're doing the same search on the site itself, this extension can't improve those results. Not all sites well designed to be searched.

Here's a [30s Video Demo](https://ext.runcode.run/rtfm/readme/RTFM_how_to.mp4)
if you prefer to WTFM.

![How to use RTFM in 3 steps](https://ext.runcode.run/rtfm/readme/RTFM_how_to.png)

## Usage

RTFM can be used either by:
- right-clicking on a term and choosing the RTFM command from the pop-up menu, OR
- selecting a word and then choosing the menu item **Editor → RTFM** (or assign a hot-key to it!) OR
- Triggering **RTFM** with no text selected to be prompted for a keyword to search for.

### Configuration

To configure the global preferences, open **Extensions → Extension Library...** then select RTFM's **Preferences** tab.

### Preferences

For each syntax, you can specify to which online resource you would like to send your searches. Searches that are done "...via Duck Duck Go redirect" use the "I'm feeling Ducky" option to automatically redirect the search to the first search result. This is *usually* the page you're
looking for, but ymmv.

#### Custom Language Resources

You can also add up to 3 Custom URLs in the "Custom Language Resources" area. This can be extremely useful if you need to target a particular library API. e.g., The Javascript settings include presets targeting the jQuery, HighCharts and Nova APIs.

URLs should start with http or https, and should include the %@ token at the location where the search string should appear. URLs can also include the %$ token where you'd like to specify the document's current syntax (helpful if you are targeting are website like Mozilla.org where they
support many syntaxes).

If the URL you enter works in your browser (except for replacing %@ and %$ tokens) it should work in RTFM.

Precisely how to create a custom search url for a website is outside the scope of this manual so you'll need to RTFM about that somewhere else. :-) But if you want to target a specific website through a search engine like Duck Duck Go, the URL generally looks like the placeholder text:

    https://duckduckgo.com/?q=\\+site:EXAMPLE.COM+%$+%@

If you include %$ in the URL template, it will be replaced with the current document syntax. The %@ token will be replaced with whatever you select in your editor to search for (or enter via the dialog).

**NOTE:** The following syntax names will not be included in the search URL even if the URL template includes the %$ token:
   'text','diff','json','ini','markdown','shell'
The syntax names are too generic to be meaningful (or possibly misleading) in most searches.

If you're using Duck Duck Go and want to trigger their "I'm Feeling Ducky" (redirect-to-first-result) feature, start the query with backslash \ . For Google you must include '&btnI' after the search term.

#### Why are you using these particular presets?

Tradeoffs. Some websites (php.net) do best when searched directly, some do better when searched through a search engine (HTML5Doctor), and some don't seem to support search at all (Vue).

I've tried to provide a decent set of reference sites without overwhelming the menu with choices for each syntax. (SQL has the most, of course, since SQL isn't just one language.)

The "I'm feeling Ducky" searches through Duck Duck Go are generally great and it's an extremely fast way to get right to the page you need. And sometimes the search engines reveal better content than the site itself.

e.g., HTML5Doctor will show you a simple definition for \<figure\> if you search for that term directly on their site, but using Duck Duck Go you get a longer article on *why* and *when* you should use \<figure\>.

Others (e.g., Perl) seem to do worse using DDG, so while it's an option, it's not the default. And some syntaxes are largely symbolic (e.g., Markdown) and it seemed like if anyone needed resources for these it would be for the content they are writing, not for the syntax of the content.

I am not familiar with every language so the presets chosen may not be ideal. If you find something better, you can create and assign you own custom URLs to any syntax, or send it to me and if works better than what's in RTFM now, I'll probably include it! (There's only 3 Custom URLs because I'm hoping that's enough for most people and I didn't want to clutter up the menus.)


CHANGELOG.md
## Version 2.0.0
2022-11-08
- THIS RELEASE IS FOR NOVA 10.1+. (Nova 10.0 had a bug that could cause problems for this version of the extension. Syntaxes have also been changed to match those in v10+.)
- Removed syntaxes no longer supported directly in the Syntax menu in Nova 10: CoffeeScript, ERB, HAML, JSX, Smarty, Vue.
- Added support for new syntaxes in the Syntax menu in Nova 10: HTML+EJS, Regex, Tree Sitter
- JSX-specific search options moved to Javascript syntax.

## Version 1.2.0
2022-09-09
- Localize for French and Spanish.
- Add Wordpress API to PHP options.
- Added JSON to the list of "generic" syntaxes that don't include their name in searches when the search URL uses %$.
- Minor change to rc_log().

## Version 1.1.1
2022-07-25
- Fix: 'undefined is not an object' error reported to the console when showInputPalette() was dismissed (clicking away or hitting escape) instead of entering text.

## Version 1.1.0
2022-07-13
- Update text-prompt dialog to use workspace.showInputPalette instead of NotificationRequest.
- Updated icon.

## Version 1.0.2
2022-06-23
- Typos. Grrrr.

## Version 1.0.1
2022-06-23
- Add jQuery and HighCharts API reference to Javascript.
- Revise rc_log();
- Wording changes to documentation.
- Updated code comments.

## Version 1.0
2022-06-22 Initial release



Scripts/main.js
// ============================================================================================== //
// RTFM - An extension for Panic's Nova IDE
// Copyright (c) 2022, Colin Foster
// LICENSE: BSD 3-Clause License
// ============================================================================================== //

exports.activate = function() {
	// Do work when the extension is activated
	nova.commands.register('rtfm.search', handle_search );
}

exports.deactivate = function() {
	// Clean up state before the extension is deactivated
}


// ---------------------------------------------------------------------------------------------- //
/*	handle_search
	
	Entry point when the RTFM command is triggered. Checks to see if the user has text highlighted,
	and if not, prompts for search_text. Based on:
	- the syntax of the current document and
	- the user's preference of search-resource for this syntax
	an appropriate web page is opened to search for the search_text.
*/
// ---------------------------------------------------------------------------------------------- //

function handle_search( editor )
{
	try
	{
		// Retrieve details about the current syntax and its search target.
		// e.g., 'css' returns [id: 'css', name: "CSS", 'resource_name': "Mozilla.org via Duck Duck Go", 'url_template': "https://..."]
		var syntax      = get_syntax_profile( editor.document.syntax );
		var search_text = editor.selectedText.trim();
		
		
		// If the user didn't have text highlighted in the editor, ask them for a string.
		if( !search_text )
		{
			nova.workspace.showInputPalette(
				_l('_prompt.search_string', syntax['resource_name'] ),
				{'placeholder': _l('_prompt.placeholder', syntax['name'] ) },
				( search_text ) =>
				{
					do_search( syntax['url_template'], syntax['id'], search_text );
				}
			);
			
			return;
		}
		
		do_search( syntax['url_template'], syntax['id'], search_text );
	}
	catch( error )
	{
		rc_log( error );
		nova_notify( _l('_err.error'), error.message );
	}
}


// ---------------------------------------------------------------------------------------------- //
/*	do_search
	
	Passed:
	- the URL template (with %@ and possibly %$ token placeholders),
	- syntax id (e.g., 'javascript', 'html'), and
	- the user's search_text,
	it substitutes the search_text (and maybe syntax) into the URL, then opens the resulting
	search_url in the user's default browser.
*/
// ---------------------------------------------------------------------------------------------- //

function do_search( url_template, syntax_id, search_text )
{
	if( rc_type( search_text ) !== 'string' || !search_text.trim().length )
		return;
	
	// Don't include generic syntax keywords for these types of searches.
	if( ['text','diff','ini','json','markdown','shell'].indexOf( syntax_id ) >= 0 )
		syntax_id = '';
	
	var search_url = url_template.replace('%@', search_text.trim() )
							.replace('%$', syntax_id );
	
	nova.openURL( encodeURI( search_url ), ( success ) =>
	{
		if( !success )
			throw new Error( _l('_err.open_url_failed') );
	});
}


// ---------------------------------------------------------------------------------------------- //
/*	get_url_template
	
	Returns the URL template for a particular syntax (e.g., javascript) and resource_id (e.g., moz).
	If the user has specified one of the Custom URLs, that text is loaded, otherwise the template
	is loaded from the presets / localization file via lookup_preset_url_template().
*/
// ---------------------------------------------------------------------------------------------- //

function get_url_template( syntax, resource_id )
{
	var url_template = false;
	if( resource_id.slice( 0, 6 ) !== 'custom')
		return lookup_preset_url_template( syntax, resource_id );
	
	// If the user has selected to use their own URL, load it and validate it.
	url_template = nova.config.get('rtfm.' + resource_id + '_resource');
	url_template = validate_url_template( url_template );
	
	if( !url_template )
		throw new Error( _l('_err.invalid_custom_url', syntax ) );
	
	return url_template;
}


// ---------------------------------------------------------------------------------------------- //
/*	validate_url_template
	
	Make sure what we've gotten from the user bears at least a passing resemblance to a valid
	search URL.
*/
// ---------------------------------------------------------------------------------------------- //

function validate_url_template( url )
{
	if( rc_type( url ) !== 'string' || url.length < 16 )
		return false;
	
	url = url.match(/^https?:\/\/[^.\/\s]+\.\S*$/);
	if( url )
		return url[ 0 ];
	
	return false;
}


// ---------------------------------------------------------------------------------------------- //
/*	get_syntax_profile
	
	Given a syntax code from the editor, it returns a structure with the key values the
	extension needs to perform the search and display any dialogs.
	
	e.g., Given 'javascript' it might return:
	{
		'id': 'javascript',
		'name': Javascript',
		'resource_name': "Mozilla.org via Duck Duck Go redirect",
		'url_template': "https://duckduckgo.com/?q=\+site:developer.mozilla.org+%@"
	}
*/
// ---------------------------------------------------------------------------------------------- //

function get_syntax_profile( syntax )
{
	if( !syntax )
		syntax = 'text';
	else if( syntax.slice( 0, 10 ) === 'typescript' )	// All typescript variations are treated the same.
		syntax = 'typescript';
	
	var resource_id = get_resource_id_from_prefs( syntax );
	
	// Check to see if we're mapping from one syntax to another. (e.g., from CoffeeScript to Javascript settings.)
	// Remapping can happen only once (the prefs are set up such that all remaps are 'end points').
	if( resource_id.slice( 0, 4 ) === 'use-' )
	{
		syntax        = resource_id.slice( 4 );	// strip off 'use-' from to get the syntax to use.
		resource_id   = get_resource_id_from_prefs( syntax );
	}
	
	var syntax_name = _l('_syntax.' + syntax );				// e.g., 'javascript' => "Javascript", 'text' => "Plain Text"
	if( resource_id === '-' )
		throw new Error( _l("_err.no_search_resource", syntax_name ) );
	
	var url_template  = get_url_template( syntax, resource_id );
			
	var resource_name = _l('_rsrc_name.' + resource_id );	// e.g., "Mozilla.org via Duck Duck Go redirect"
	if( resource_id === 'custom')
		resource_name += ' (' + url_template.split('/')[ 2 ] + ')';	// Tack on the domain of Custom URL.
	
	var syntax_profile =
	{
		'id': syntax,
		'name': syntax_name,
		'resource_name': resource_name,
		'url_template': url_template
	};
	
	return syntax_profile;
}


// ---------------------------------------------------------------------------------------------- //
/*	lookup_preset_url_template
	
	Returns the search URL template, containing %$ and %@ placeholder tokens for the search terms.
	
	URL templates are localized in strings.json so it's possible to target websites in languages
	other than English, if they support them.
*/
// ---------------------------------------------------------------------------------------------- //

function lookup_preset_url_template( syntax, resource_id )
{	// General searches to Duck Duck Go and Google (not site: target),
	// use _general not their syntax in the key.
	var url_localization_key = ( ['ddg','goog'].indexOf( resource_id ) > -1 )
		? '_url._general.' + resource_id
		: '_url.' + syntax + '.' + resource_id;
	
	url_template = _l( url_localization_key );
	
	// If localization failed / doesn't start http, throw an error.
	if( !url_template || url_template.slice( 0, 4 ) !== 'http')
		throw new Error("Unable to find valid preset url template.\n[" + syntax + "," + resource_id + "]");
	
	return url_template;
}


// ---------------------------------------------------------------------------------------------- //
/*	get_resource_id_from_prefs
*/
// ---------------------------------------------------------------------------------------------- //

function get_resource_id_from_prefs( syntax )
{
	var resource_id = nova.config.get('rtfm.' + syntax + '_resource');
	if( !resource_id )
		throw new Error( _l('_err.cant_load_prefs', syntax ) );
	
	return resource_id;
}


// ---------------------------------------------------------------------------------------------- //
/*	nova_notify
	
	Show a standard notification prompt
*/
// ---------------------------------------------------------------------------------------------- //

function nova_notify( title, body, duration )
{
	var fn        = nova_notify;
	var note_id   = 'rtfm-notify';
	
	if( typeof( duration ) === 'undefined' )
		duration = 8000;
	
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -//
	/*	show
	*/
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -//
	
	fn.show = function()
	{
		fn.hide();		// Hide any previous notification we posted.
		
		var request     = new NotificationRequest( note_id );
		
		request.title   = _l( title );
		request.body    = _l( body );
		request.actions = [ _l('_prompt.dismiss') ];
		
		nova.notifications.add( request );
		
		if( duration )
			fn[ note_id ] = setTimeout( fn.hide, duration );
	}
	
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -//
	/*	hide
	*/
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -//

	fn.hide = function()
	{	// NOTE: fn[ note_id ] may contain a valid 'timeout' id of 0.
		if( typeof( fn[ note_id ] ) === 'undefined' || fn[ note_id ] === false )
			return;
		
		clearTimeout( fn[ note_id ] );
		nova.notifications.cancel( note_id );
		fn[ note_id ] = false;
	}
	
	// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -//
	
	fn.show();
}


// ---------------------------------------------------------------------------------------------- //
/*	_l - Localization

	Takes any number of arguments following the key which are replaced in the requested string.
	The first argument after the key replaces '%@1', the 2nd replaces '%@2', etc.
*/
// ---------------------------------------------------------------------------------------------- //

function _l( key )
{
	var str = nova.localize( key );
	
	// Arguments are replaced in reverse order so we replace %@11 (if it exists) before %@1.
	for( var i = arguments.length - 1; i > 0; i-- )
		str = str.replace('%@'+i, nova.localize( arguments[ i ] ) );
	
	return str;
}


// ============================================================================================== //
// LIBRARY ITEMS
// ============================================================================================== //

// ---------------------------------------------------------------------------------------------- //
/*	rc_type - v1.0.0 2022-06-21
	
	Does all the types typeof does, plus:
	'null', 'undefined','array', 'error', 'date', 'set', 'map', 'object', and any other class name.
*/
// ---------------------------------------------------------------------------------------------- //

function rc_type( o )
{
	return Object.prototype.toString.call(o).slice( 8, -1 ).toLowerCase();
}


// ---------------------------------------------------------------------------------------------- //
/*	rc_log - v2.1.1 2022-08-11
*/
// ---------------------------------------------------------------------------------------------- //

function rc_log()
{
	rc_log.id++;
	
	var log_type_and_value = function( type, value, prefix )
	{
		value = ['null','undefined'].indexOf( type ) === -1 ? " (" + value.toString() + ")" : '';
		
		console.log( prefix + type + value );
	};
	
	try{ var caller_name = arguments.callee.caller.name ? arguments.callee.caller.name : 'anon'; }
	catch( e ){ caller_name = 'unknown'; }
	
	console.log( "= " + rc_log.id + " =======================  " + caller_name + "()" );
		
	for( var i = 0; i < arguments.length; i++ )
	{
		var o = arguments[ i ];
		var type = rc_type( o );
		
		if( ['array','object'].indexOf( type ) > -1 )
		{
			var keys = Object.keys( o );
			if( !keys.length )
				log_type_and_value( type, o, "- " );
			else
			{
				console.log('{');
				keys.forEach( function( key )
				{
					log_type_and_value( rc_type( o[ key ] ), o[ key ], '   "' + key + '": ' );
				} );
				console.log('}');
			}
		}
		else if( type === 'error')
			console.log( "X Line " + o.line + ":" + o.column + " " + o.message );
		else
			log_type_and_value( type, o, "- " );
	}
}
rc_log.id = 0;
dbg = rc_log;	// To differentiate between temporary and permanent logging.


// ---------------------------------------------------------------------------------------------- //
